storage/src/adb/any/variable/sync.rs

```diff
@@ -213,16 +498,224 @@ async fn compute_offset<E: Storage + Metrics, V: Codec>(
+    lower_bound: u64,
+    upper_bound: u64,
+    mut oldest_retained_loc: u64,
+    items_per_section: NonZeroU64,
+) -> Result<(u64, u64), crate::journal::Error>
+where
+    E: Storage + Metrics + Clock,
+    V: Codec,
+{
+    if lower_bound > upper_bound {
+        return Err(crate::journal::Error::InvalidSyncRange(
+            lower_bound,
+            upper_bound,
+        ));
+    }
+
+    let items_per_section_val = items_per_section.get();
+    let lower_section = lower_bound / items_per_section_val;
+    let upper_section = upper_bound / items_per_section_val;
+
+    debug!(
+        lower_bound,
+        upper_bound,
+        oldest_retained_loc,
+        lower_section,
+        upper_section,
+        items_per_section = items_per_section_val,
+        "pruning journal"
+    );
+
+    // Remove sections before the lower_section
+    if lower_section > 0 {
+        journal.prune(lower_section).await?;
+    }
+    // If the oldest_retained_loc was just pruned, we need to update it
+    if oldest_retained_loc < lower_section * items_per_section_val {
+        oldest_retained_loc = lower_section * items_per_section_val;
+    }
+
+    // Remove sections after the upper_section
+    let sections_to_remove: Vec<u64> = journal
+        .blobs
+        .range((Bound::Excluded(upper_section), Bound::Unbounded))
+        .map(|(&section, _)| section)
+        .collect();
+
+    for section in sections_to_remove {
+        debug!(section, "removing section beyond upper bound");
+        if let Some(blob) = journal.blobs.remove(&section) {
+            drop(blob);
+            let name = section.to_be_bytes();
+            journal
+                .context
+                .remove(&journal.cfg.partition, Some(&name))
+                .await?;
+            journal.tracked.dec();
+        }
+    }
+
+    // Prune the lower section if needed
+    let lower_section_start = lower_section * items_per_section_val;
+    if lower_bound > lower_section_start && journal.blobs.contains_key(&lower_section) {
+        prune_lower(journal, lower_bound, items_per_section, oldest_retained_loc).await?;
+    }
+
+    // Prune the upper section if needed
+    prune_upper(journal, upper_bound, items_per_section_val).await?;
+
+    // Step 5: Compute the next location to write and the new oldest_retained_loc
+    // If journal is empty, return (lower_bound, lower_bound)
+    if journal.blobs.is_empty() {
+        return Ok((lower_bound, lower_bound));
+    }
+
+    // If oldest_retained_loc was before lower_bound, we've now pruned up to lower_bound.
+    oldest_retained_loc = lower_bound.max(oldest_retained_loc);
+
+    let last_section_start = journal
+        .blobs
+        .last_key_value()
+        .map(|(section, _)| section * items_per_section_val)
+        .unwrap();
+    let last_section_end = last_section_start + items_per_section_val - 1;
+    let next_loc = (last_section_end + 1).min(upper_bound + 1);
+
+    Ok((next_loc, oldest_retained_loc))
```

Comment:
This `next_loc` calculation is subtly wrong in some way. This [diff](https://github.com/commonwarexyz/monorepo/pull/1433/commits/8c6fd06f0ff3356fa3a8c9fae15c6fc452c3c38c#diff-d3d98e6170e0e8e97154acd6fd56f2d7d15436d4d7ee48da5525bf196fbae9e7R679) causes the sync example client to crash after the first iteration. I need to figure out where this bug is before we merge this.